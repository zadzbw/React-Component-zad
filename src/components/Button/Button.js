import React, {PropTypes, Component}from 'react';
import classNames from 'classnames';
import oneOf from '../../utils/oneOf';

import './Button.less';

const _type = {
	type: ['primary', 'ghost', 'dashed'],
	size: ['small', 'large'],
	shape: ['circle', 'circle-o']
};
const btnPrefix = 'zad-btn';

export default class Button extends Component {
	constructor(props) {
		super(props);
	}

	static propTypes = {
		type: PropTypes.string,
		size: PropTypes.string,
		shape: PropTypes.string,
		onClick: PropTypes.func
	};

	static defaultProps = {
		onClick: () => undefined
	};


	render() {
		const {type, size, shape, className, children, ...props} = this.props;

		const sizeClass = ({
				large: 'lg',
				small: 'sm',
			})[size] || '';

		const btnClass = classNames(btnPrefix, className, {
			[`${btnPrefix}-${type}`]: type && oneOf(_type.type, type),
			[`${btnPrefix}-${sizeClass}`]: size && oneOf(_type.size, size),
			[`${btnPrefix}-${shape}`]: shape && oneOf(_type.shape, shape)
		});

		return (
			<button
				{...props}
				type="button"
				className={btnClass}
			>
				<span>{children}</span>
			</button>
		);
	}
}
