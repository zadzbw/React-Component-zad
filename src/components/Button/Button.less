@import "../../style/variables";
@import "./ButtonMixin";

.zad-btn {
  .btn();
  .btn-default();

  > .fa + span, > span + .fa, > .fa + .fa {
    margin-left: 0.5em;
  }

  &-primary {
    .btn-primary();
  }

  &-ghost {
    .btn-ghost();
  }

  &-dashed {
    .btn-dashed();
  }

  &-circle {
    .btn-circle();
  }

  &-clicked:before {
    content: '';
    position: absolute;
    border-radius: inherit;
    border: 0 solid @primary-color;
    animation: buttonClick @animation-duration-slow ease-out forwards;
  }
}

@keyframes buttonClick {
  from {
    opacity: 1;
    top: -1px;
    left: -1px;
    bottom: -1px;
    right: -1px;
    border-width: 0;
  }

  to {
    opacity: 0;
    top: -6px;
    left: -6px;
    bottom: -6px;
    right: -6px;
    border-width: 6px;
  }
}