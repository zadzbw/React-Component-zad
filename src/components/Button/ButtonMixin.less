@import "../../style/variables";

.btn-size(@padding;@font-size;@border-radius) {
  padding: @padding;
  font-size: @font-size;
  border-radius: @border-radius;
}

.btn-color(@color;@background;@border) {
  color: @color;
  background-color: @background;
  border-color: @border;
}

.btn-disabled() {
  &[disabled] {
    cursor: not-allowed;
    > * {
      pointer-events: none;
    }

    &, &:hover, &:focus, &:active {
      .btn-color(@btn-disable-color; @btn-disable-bg; @btn-disable-border);
    }
  }
}

.btn-variant-primary(@color;@background) {
  .btn-color(@color; @background; @background);

  &:hover, &:focus {
    .btn-color(@color; tint(@background, 20%); tint(@background, 20%));
  }

  &:active {
    .btn-color(@color; shade(@background, 5%); shade(@background, 5%));
  }

  .btn-disabled();
}

.btn-variant-others(@color;@background;@border) {
  .btn-color(@color; @background; @border);

  &:hover, &:focus {
    .btn-color(tint(@primary-color, 20%); @background; tint(@primary-color, 20%));
  }

  &:active {
    .btn-color(shade(@primary-color, 5%); @background; shade(@primary-color, 5%));
  }

  .btn-disabled();
}

.btn() {
  position: relative;
  display: inline-block;
  text-align: center;
  font-weight: @btn-font-weight;
  line-height: @line-height-base;
  transition: all @animation-duration-base @ease-in-out;
  margin-bottom: 0;
  border: 1px solid transparent;
  touch-action: manipulation;
  cursor: pointer;
  background-image: none;
  white-space: nowrap;
  user-select: none;

  &, &:active, &:focus {
    outline: 0;
  }

  &:not([disabled]):hover {
    text-decoration: none;
  }

  &:not([disabled]):active {
    outline: 0;
    transition: none;
  }

  .btn-size(@btn-padding-base; @font-size-base; @btn-border-radius-base);

  &-sm {
    .btn-size(@btn-group-padding-sm; @font-size-sm; @btn-border-radius-sm);
  }

  &-lg {
    .btn-size(@btn-padding-lg; @btn-font-size-lg; @btn-border-radius-base);
  }
}

.btn-default() {
  &:hover, &:focus, &:active, &.active {
    background: #fff;
  }
  .btn-variant-others(@btn-default-color; @btn-default-bg; @btn-default-border);
}

.btn-primary() {
  .btn-variant-primary(@btn-primary-color; @btn-primary-bg);
}

.btn-ghost() {
  .btn-variant-others(@btn-ghost-color, @btn-ghost-bg, @btn-ghost-border);
}

.btn-dashed() {
  .btn-ghost();
  border-style: dashed;
}

.btn-circle() {
  top: 1px; // 清除circle向上偏移的1px
  width: @btn-circle-size-base;
  height: @btn-circle-size-base;
  .btn-size(0; @font-size-base + 2; 50%);

  &.zad-btn-sm {
    top: 0;
    width: @btn-circle-size-sm;
    height: @btn-circle-size-sm;
    .btn-size(0; @font-size-sm + 2; 50%);
  }

  &.zad-btn-lg {
    top: 0;
    width: @btn-circle-size-lg;
    height: @btn-circle-size-lg;
    .btn-size(0; @btn-font-size-lg + 2; 50%);
  }
}
