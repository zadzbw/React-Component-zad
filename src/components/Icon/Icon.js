import React, {PropTypes, Component} from 'react';
import classNames from 'classnames';
import isOneOf from '../../utils/isOneOf';

const _type = {
  size: ['lg', '2x', '3x', '4x', '5x'],
  rotate: [90, 180, 270]
};

export default class Icon extends Component {
  static propTypes = {
    name: PropTypes.string.isRequired,
    size: PropTypes.string,
    fixedWidth: PropTypes.bool,
    spin: PropTypes.bool,
    pulse: PropTypes.bool,
    rotate: PropTypes.number
  };

  render() {
    const {name, size, fixedWidth, spin, pulse, rotate, ...props} = this.props;
    const iconClass = classNames('fa', `fa-${name}`, {
      [`fa-${size}`]: size && isOneOf(_type.size, size),
      [`fa-rotate-${rotate}`]: rotate && isOneOf(_type.rotate, rotate),
      'fa-fw': fixedWidth,
      'fa-spin': spin,
      'fa-pulse': pulse
    });
    return (
      <i {...props} className={iconClass}/>
    );
  }
}
