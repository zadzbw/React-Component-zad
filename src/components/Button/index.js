/**
 * Created by zad on 16/11/30.
 */
import Button from './Button';

export default Button;