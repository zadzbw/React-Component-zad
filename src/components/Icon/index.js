/**
 * Created by zad on 16/11/30.
 */
import Icon from './Icon';

export default Icon;