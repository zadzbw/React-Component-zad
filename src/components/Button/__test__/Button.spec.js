/**
 * Created by zad on 16/12/7.
 */
import React from 'react';
import {mount} from 'enzyme';
import {mountToJson} from 'enzyme-to-json';
import Button from '../../Button';
import Icon from '../../Icon';

describe('Button test', function () {
  it('type', () => {
    const wrapper = mount(
      <Button type={'primary'}>primary</Button>
    );
    expect(wrapper.find('button').type()).toEqual('button');
    expect(wrapper.find('button').text()).toEqual('primary');
    expect(wrapper.find('button').hasClass('zad-btn-primary')).toBeTruthy();
    expect(mountToJson(wrapper)).toMatchSnapshot();
  });

  it('size', () => {
    const wrapper = mount(
      <Button size={'large'}>large</Button>
    );
    expect(wrapper.find('button').type()).toEqual('button');
    expect(wrapper.find('button').text()).toEqual('large');
    expect(wrapper.find('button').hasClass('zad-btn-lg')).toBeTruthy();
    expect(mountToJson(wrapper)).toMatchSnapshot();
  });

  it('shape', () => {
    const wrapper = mount(
      <Button shape={'circle'}>circle</Button>
    );
    expect(wrapper.find('button').type()).toEqual('button');
    expect(wrapper.find('button').text()).toEqual('circle');
    expect(wrapper.find('button').hasClass('zad-btn-circle')).toBeTruthy();
    expect(mountToJson(wrapper)).toMatchSnapshot();
  });

  it('clicked', () => {
    let a = 1;

    function test() {
      a++;
    }

    const wrapper = mount(
      <Button onClick={test}>button</Button>
    );
    expect(a).toEqual(1);
    wrapper.find('button').simulate('click');
    expect(a).toEqual(2);
    expect(mountToJson(wrapper)).toMatchSnapshot();
  });

  it('disabled', () => {
    const wrapper = mount(
      <Button disabled>button</Button>
    );

    expect(wrapper.find('button').get(0).disabled).toBeTruthy();
    expect(wrapper.find('button').type()).toEqual('button');
    expect(wrapper.find('button').text()).toEqual('button');
    expect(wrapper.find('button').hasClass('zad-btn')).toBeTruthy();
    expect(mountToJson(wrapper)).toMatchSnapshot();
  });

  it('Icon', () => {
    const wrapper = mount(
      <Button disabled><Icon name={'shield'}/>hello</Button>
    );

    expect(wrapper.find('i').hasClass('fa-shield')).toBeTruthy();
    expect(wrapper.find('button').text()).toEqual('hello');
    expect(mountToJson(wrapper)).toMatchSnapshot();
  });
});