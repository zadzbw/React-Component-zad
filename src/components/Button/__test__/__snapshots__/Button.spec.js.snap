exports[`Button test Icon 1`] = `
<Button
  disabled={true}
  onClick={[Function]}>
  <button
    className="zad-btn"
    disabled={true}
    onClick={[Function]}
    onMouseUp={[Function]}
    type="button">
    <Icon
      name="shield">
      <i
        className="fa fa-shield" />
    </Icon>
    <span>
      hello
    </span>
  </button>
</Button>
`;

exports[`Button test clicked 1`] = `
<Button
  onClick={[Function]}>
  <button
    className="zad-btn"
    onClick={[Function]}
    onMouseUp={[Function]}
    type="button">
    <span>
      button
    </span>
  </button>
</Button>
`;

exports[`Button test disabled 1`] = `
<Button
  disabled={true}
  onClick={[Function]}>
  <button
    className="zad-btn"
    disabled={true}
    onClick={[Function]}
    onMouseUp={[Function]}
    type="button">
    <span>
      button
    </span>
  </button>
</Button>
`;

exports[`Button test shape 1`] = `
<Button
  onClick={[Function]}
  shape="circle">
  <button
    className="zad-btn zad-btn-circle"
    onClick={[Function]}
    onMouseUp={[Function]}
    type="button">
    <span>
      circle
    </span>
  </button>
</Button>
`;

exports[`Button test size 1`] = `
<Button
  onClick={[Function]}
  size="large">
  <button
    className="zad-btn zad-btn-lg"
    onClick={[Function]}
    onMouseUp={[Function]}
    type="button">
    <span>
      large
    </span>
  </button>
</Button>
`;

exports[`Button test type 1`] = `
<Button
  onClick={[Function]}
  type="primary">
  <button
    className="zad-btn zad-btn-primary"
    onClick={[Function]}
    onMouseUp={[Function]}
    type="button">
    <span>
      primary
    </span>
  </button>
</Button>
`;
